README.md
# jsun

This is an alternative JSON decoder/encoder that supports some extra
features. It takes a *lot* of inspiration from the `json` module in the
standard library and exposes the same high level API: `load`, `loads`,
`dump`, `dumps`, `JSONDecoder`, and `JSONEncoder`.

In many cases, `jsun` can be swapped in by installing the package then
simply updating imports to use `jsun` instead of `json`.

## Extra decoding features

- Trailing commas
 
- Line comments starting with //
 
- All valid Python ints and floats:
  - Binary, octal, hex
  - Underscore separators
  - Unary plus operator

- Math constants:
  - inf, nan, E, π, PI, τ, TAU
  - Infinity, NaN

- Literal (unquoted) dates and times:
  - 2021-06
  - 2021-06-23
  - 2021-06-23T12:00
  - 2021-06-23T12:00Z
  - 2021-06-23T12:00-07:00
  - 12:00 (today's date at noon)

- Decoding an empty string will produce `None` rather than an exception
  (an exception will be raised if extras are disabled)

- *All* parsing methods can be overridden if some additional
  customization is required. In particular, the object and array
  parsers can be overridden

- A pre-parse method can be provided to handle values before the regular
  JSON parsers are applied

- A fallback parsing method can be provided to handle additional types
  of values if none of the default parsers are suitable

- When errors are encountered, specific exceptions are raised (all
  derived from the built-in `ValueError`)

## Extra encoding features

The `jsun` encoder is very similar to the standard library encoder (and
is in fact a subclass of `json.JSONEncoder`). Currently, it supports
only a couple of extra features:

- Date objects are converted to ISO format by default
- Datetime objects are converted to ISO format by default

NOTE: There is some asymmetry here. E.g., date and datetime objects
should be converted to literals instead of quoted strings.

## Disabling the extra features

*All* the extra features can be turned off with a flag:

    >>> from jsun import decode
    >>> decode("[1, 2, 3,]")
    [1, 2, 3]
    >>> decode("[1, 2, 3,]", enable_extras=False)
    <exception traceback>

## Differences between jsun and standard library json

- An empty string input is converted to `None` rather than raising an
  exception (only if extras are enabled).

- When decoding, instead of `object_hook` and `object_hook_pairs`,
  there's just a single `object_converter` argument. It's essentially
  the same as `object_hook`. `object_hook_pairs` seems unnecessary
  nowadays since `dict`s are ordered.

- The default object type is `jsun.obj.JSONObject` instead of `dict`. A
  `JSONObject` is a bucket of properties that can be accessed via dotted
  or bracket notation. Pass `object_converter=None` to get back
  `dict`s instead.

## Config files

A bonus feature is that configuration can be loaded from INI files
where the keys are split on dots to create sub-objects and the values
are encoded as JSON.

This is quite similar to TOML and some of the features of `jsun`, like
literal dates, are inspired by TOML.

This feature was originally developed in 2014 as part of the
`django-local-settings` project, about a year and half after TOML was
first released but before I'd heard of it.

### Differences with TOML

- Parentheses are used instead of quotes to avoid splitting keys on dots
- Objects created using `{}` syntax (AKA "inline tables" in TOML) can
  span multiple lines
- There are no arrays of tables
- Others I'm not thinking of at the moment...

## About the name

My first choice was `jsonish` but that's already taken. My second choice
was `jsonesque` but it's also taken, and it's hard to type. `jsun` is
nice because it's easy to type and easy to swap in for `json` by just
changing a single letter.

## Testing

There's a suite of unit tests, which also tests against the JSON checker
files at https://json.org/JSON_checker/. Coverage is currently at 82%.


CHANGELOG.md
# Change log for jsun

## 1.0 - unreleased

In progress...

## 1.0a1 - 2021-11-03

This is the initial standalone version of this package. It was spun out
of the [https://github.com/wylee/django-local-settings](django-local-settings)
package.


tests/test_decoder.py
import doctest
import math
import unittest
from pathlib import Path

import jsun.decoder
import jsun.scanner

from jsun import decode, decode_file
from jsun.exc import (
    DecodeError,
    ExpectedKeyError,
    ExpectedValueError,
    ExtraneousDataError,
    ScanStringError,
    UnexpectedCharError,
    UnknownCharError,
    UnmatchedBracketError,
    INIDecodeError,
)


def load_tests(loader, tests, ignore):
    tests.addTests(doctest.DocTestSuite(jsun.decoder))
    tests.addTests(doctest.DocTestSuite(jsun.obj))
    tests.addTests(doctest.DocTestSuite(jsun.scanner))
    return tests


class TestJSONishScanner(unittest.TestCase):
    def decode(self, string, object_converter=None, enable_extras=True):
        return decode(
            string,
            object_converter=object_converter,
            enable_extras=enable_extras,
        )

    def test_empty_string_is_none(self):
        result = self.decode("")
        self.assertIsNone(result)
        self.assertRaises(ExpectedValueError, self.decode, " ")

    def test_inf(self):
        self.assertEqual(self.decode("inf"), math.inf)
        self.assertEqual(self.decode("+inf"), math.inf)
        self.assertEqual(self.decode("-inf"), -math.inf)

    def test_nan(self):
        self.assertTrue(math.isnan(float("nan")))
        self.assertTrue(math.isnan(self.decode("nan")))
        self.assertTrue(math.isnan(self.decode("+nan")))
        self.assertTrue(math.isnan(self.decode("-nan")))

    def test_empty_object(self):
        self.assertEqual(self.decode("{}"), {})

    def test_empty_string(self):
        self.assertIsNone(self.decode(""))
        self.assertRaises(ExpectedValueError, self.decode, "", enable_extras=False)

    def test_decode_multiline_object(self):
        self.assertEqual(self.decode("{\n\n\n}"), {})

    def test_empty_array(self):
        self.assertEqual(self.decode("[]"), [])

    def test_decode_multiline_array(self):
        self.assertEqual(self.decode("[\n\n\n]"), [])

    def test_simple_object(self):
        self.assertEqual(self.decode('{"a": 1}'), {"a": 1})

    def test_object_with_space_before_key(self):
        self.assertEqual(self.decode('{ "a": 1}'), {"a": 1})

    def test_unclosed_object(self):
        self.assertRaises(UnmatchedBracketError, self.decode, '{"a": 1')
        self.assertRaises(ExtraneousDataError, self.decode, '"a": 1}')

    def test_unclosed_array(self):
        self.assertRaises(UnmatchedBracketError, self.decode, "[1, 2")
        self.assertRaises(ExtraneousDataError, self.decode, "1, 2]")

    def test_nesting(self):
        result = self.decode('{ "1": 0b1, "2": [[1, 2]], }')
        self.assertEqual(result, {"1": 1, "2": [[1, 2]]})

    def test_comments(self):
        self.decode(
            """
            // This is a JSON object
            {
                // "a" is really special
                "a": 1,
                "b": 2,  // end-of-line comment
            }
            """
        )
        self.assertEqual(self.decode('"//"'), "//")
        self.assertRaises(ExpectedValueError, self.decode, "//{}")
        self.assertRaises(ExpectedKeyError, self.decode, "{//}")
        self.assertRaises(ExpectedValueError, self.decode, '{"a": //}')
        self.assertRaises(UnmatchedBracketError, self.decode, '{"a": 1//}')

    def test_comments_with_extra_features_disabled(self):
        doc = """
        // Comment
        {
            // Comment
            "a": 1,
            "b": 2,
        }
        """
        self.assertRaises(UnknownCharError, self.decode, doc, enable_extras=False)

    def test_trailing_commas_with_extra_features_disabled(self):
        self.assertRaises(
            UnexpectedCharError, self.decode, "[1, 2,]", enable_extras=False
        )


class TestJSONishAgainstJSONCheckerFiles(unittest.TestCase):
    def decode_file(self, name, enable_extras=True):
        file_name = f"{name}.json"
        path = Path(__file__).parent / "json_checker_files" / file_name
        return decode_file(path, enable_extras=enable_extras)

    def test_pass1_with_extra_features_disabled(self):
        # Standard JSON shouldn't require any extra features.
        self.decode_file("pass1", enable_extras=False)

    def test_pass1_with_extra_features_enabled(self):
        # JSONish's extra features are a superset of the standard
        # features, so there shouldn't be any issues parsing a standard
        # JSON doc with them turned on.
        self.decode_file("pass1")

    def test_pass2_with_extra_features_disabled(self):
        self.decode_file("pass2", enable_extras=False)

    def test_pass2_with_extra_features_enabled(self):
        self.decode_file("pass2")

    def test_pass3_with_extra_features_disabled(self):
        self.decode_file("pass3", enable_extras=False)

    def test_pass3_with_extra_features_enabled(self):
        self.decode_file("pass3")

    def test_expected_failures(self):
        root = Path(__file__).parent / "json_checker_files"
        paths = root.glob("fail*.json")
        for path in paths:
            if path.stem in ("fail1", "fail18"):
                # These match stdlib behavior (i.e., they don't fail)
                decode_file(path)
                decode_file(path, enable_extras=False)

            elif path.stem in (
                "fail15",
                "fail17",
                "fail25",
                "fail26",
                "fail27",
                "fail28",
            ):
                self.assertRaises(
                    ScanStringError, decode_file, path, enable_extras=False
                )

            else:
                # TODO: Check for specific exceptions?
                with self.subTest(path.name):
                    self.assertRaises(
                        DecodeError, decode_file, path, enable_extras=False
                    )


class TestDecodeINI(unittest.TestCase):
    def test_decode_ini(self):
        config = """\
[section.one]
a.b = 1
x.y = 2
[section.two]
a.b = 3
x.y = 4
[other]
a = 1
b = 2
[x.(not.split)]
(x.y) = 1
(also.not.split).x = 2
        """
        result = decode(config, ini=True)
        self.assertEqual(
            result,
            {
                "section": {
                    "one": {"a": {"b": 1}, "x": {"y": 2}},
                    "two": {"a": {"b": 3}, "x": {"y": 4}},
                },
                "other": {
                    "a": 1,
                    "b": 2,
                },
                "x": {
                    "not.split": {
                        "x.y": 1,
                        "also.not.split": {
                            "x": 2,
                        },
                    },
                },
            },
        )

    def test_bad_section_names(self):
        bad_names = [
            "[section)]",
            "[(section]",
            "[x(yz)]",
            "[(xy)z]",
            "[)(abc)]",
            "[(abc)(]",
            "[(abc))]",
            "[((xyz))]",
        ]
        for name in bad_names:
            with self.subTest(name=name):
                self.assertRaises(INIDecodeError, decode, name, ini=True)


tests/test_encoder.py
import doctest

import jsun


def load_tests(loader, tests, ignore):
    tests.addTests(doctest.DocTestSuite(jsun.encoder))
    return tests


src/jsun/exc.py
class DecodeError(ValueError):
    def __init__(self, string, position, message):
        line = string.count("\n", 0, position) + 1
        column = position - string.rfind("\n", 0, position)
        super_message = (
            f"{message} at line {line} column {column} (position {position})"
        )
        super().__init__(super_message)
        self.string = string
        self.message = message
        self.position = position
        self.line = line
        self.column = column


class ExpectedBracketError(DecodeError):
    def __init__(self, string, position, bracket, message=None):
        if message is None:
            message = f"Expected bracket `{bracket}`"
        super().__init__(string, position, message)
        self.bracket = bracket


class ExpectedDelimiterError(DecodeError):
    def __init__(self, string, position, delimiter, message=None):
        if message is None:
            message = f"Expected delimiter `{delimiter}`"
        super().__init__(string, position, message)
        self.delimiter = delimiter


class ExpectedKeyError(DecodeError):
    def __init__(self, string, position, message="Expected key"):
        super().__init__(string, position, message)


class ExpectedValueError(DecodeError):
    def __init__(self, string, position, message="Expected value"):
        super().__init__(string, position, message)


class ExtraneousDataError(DecodeError):
    def __init__(
        self,
        string,
        position,
        message="Extraneous data (likely indicating a malformed JSON document)",
    ):
        super().__init__(string, position, message)


class UnexpectedCharError(DecodeError):
    def __init__(self, string, position, char, message=None):
        if message is None:
            message = f"Unexpected char `{char}`"
        super().__init__(string, position, message)
        self.char = char


class UnknownCharError(DecodeError):
    def __init__(self, string, position, char, message=None):
        if message is None:
            message = f"Unknown char `{char}`"
        super().__init__(string, position, message)
        self.char = char


class UnmatchedBracketError(DecodeError):
    def __init__(self, string, bracket, position, message=None):
        if message is None:
            message = f"Unmatched bracket `{bracket}`"
        super().__init__(string, position, message)
        self.bracket = bracket


class ScanStringError(DecodeError):
    def __init__(self, string, position, message=None):
        if message is None:
            message = "Error scanning string"
        super().__init__(string, position, message)


class INIDecodeError(DecodeError):
    def __init__(self, string, position, message):
        super_message = f"{message} at position {position}"
        super(ValueError, self).__init__(super_message)
        self.string = string
        self.message = message
        self.position = position


src/jsun/encoder.py
"""JSONish Encoder

This passes through to the stdlib `json` module with handling of
:class:`JSONObject` and :class:`datetime.datetime` values.

"""
import datetime
import json

from .obj import JSONObject


__all__ = ["encode", "encode_to_file", "Encoder"]


class Encoder(json.JSONEncoder):
    def __init__(
        self,
        *,
        skipkeys=False,
        ensure_ascii=True,
        check_circular=True,
        allow_nan=True,
        sort_keys=False,
        indent=None,
        separators=None,
        default=None,
        enable_extras=True,
    ):
        super().__init__(
            skipkeys=skipkeys,
            ensure_ascii=ensure_ascii,
            check_circular=check_circular,
            allow_nan=allow_nan,
            sort_keys=sort_keys,
            indent=indent,
            separators=separators,
            default=default,
        )
        self.enable_extras = enable_extras

    def default(self, obj):
        if self.enable_extras:
            if isinstance(obj, JSONObject):
                return dict(obj)
            # NOTE: datetime objects are also date objects
            if isinstance(obj, datetime.datetime):
                return obj.isoformat()
            if isinstance(obj, datetime.date):
                return obj.isoformat()
        # Raise TypeError
        return super().default(obj)


def encode(
    obj,
    *,
    cls=Encoder,
    skipkeys=False,
    ensure_ascii=True,
    check_circular=True,
    allow_nan=True,
    sort_keys=False,
    indent=None,
    separators=None,
    default=None,
    enable_extras=True,
):
    instance = cls(
        skipkeys=skipkeys,
        ensure_ascii=ensure_ascii,
        check_circular=check_circular,
        allow_nan=allow_nan,
        sort_keys=sort_keys,
        indent=indent,
        separators=separators,
        default=default,
        enable_extras=enable_extras,
    )
    return instance.encode(obj)


def encode_to_file(
    obj,
    fp,
    *,
    cls=Encoder,
    skipkeys=False,
    ensure_ascii=True,
    check_circular=True,
    allow_nan=True,
    sort_keys=False,
    indent=None,
    separators=None,
    default=None,
    enable_extras=True,
):
    instance = cls(
        skipkeys=skipkeys,
        ensure_ascii=ensure_ascii,
        check_circular=check_circular,
        allow_nan=allow_nan,
        sort_keys=sort_keys,
        indent=indent,
        separators=separators,
        default=default,
        enable_extras=enable_extras,
    )
    iterable = instance.iterencode(obj)
    for chunk in iterable:
        fp.write(chunk)


src/jsun/__main__.py
import argparse
import pathlib
import sys

from .decoder import decode, decode_file
from .encoder import encode, encode_to_file


def main(argv=None):
    parser = argparse.ArgumentParser(prog="jsun")

    type_group = parser.add_mutually_exclusive_group(required=True)
    type_group.add_argument("-j", "--json", help="JSON string to decode")
    type_group.add_argument("-f", "--file", help="File to decode")

    parser.add_argument(
        "-c",
        "--ini",
        action="store_true",
        default=None,
        help="Force INI+JSON",
    )

    parser.add_argument("-o", "--out-file", help="Write decoded JSON to file")

    parser.add_argument(
        "-i",
        "--indent",
        type=int,
        default=None,
        help="Pretty print with specified indent level",
    )

    parser.add_argument(
        "-d",
        "--disable-extras",
        dest="enable_extras",
        action="store_false",
        default=True,
    )

    args = parser.parse_args(argv)
    json = args.json
    file = args.file
    out_file = args.out_file
    ini = args.ini
    indent = args.indent
    enable_extras = args.enable_extras

    if json:
        obj = decode(json, enable_extras=enable_extras, ini=ini)
    elif file:
        obj = decode_file(file, enable_extras=enable_extras, ini=ini)

    if out_file:
        path = pathlib.Path(out_file)
        with path.open("w") as fp:
            encode_to_file(obj, fp, indent=indent, enable_extras=enable_extras)
    else:
        print(encode(obj, indent=indent, enable_extras=enable_extras))


if __name__ == "__main__":
    sys.exit(sys.argv)


src/jsun/decoder.py
"""JSONish Decoder

In addition to standard JSON, the JSONish decoder also supports/handles
the following:

- Trailing commas

- Line comments starting with //

- Any valid Python int or float:
  - Literal binary, octal, and hex values
  - Underscore separators in numbers
  - Unary plus operator

- Math constants:
  - inf, nan, E, π, PI, τ, TAU
  - Infinity, NaN

- Empty input strings will be converted to ``None`` rather than raising
  an exception.

- Literal (unquoted) dates & times:
  - 2021-06
  - 2021-06-23
  - 2021-06-23T12:00
  - 2021-06-23T12:00Z
  - 2021-06-23T12:00-07:00
  - 12:00 (today's date at noon)

.. note:: For dates and times, when a time zone isn't specified, the
    local time zone will be used.

- An object converter can be specified to convert JSON objects into a
  custom type (or types). By default, JSON objects will be converted to
  :class:`scanner.JSONObject`s, which allows properties to be accessed
  with either dotted or bracket notation.

- Loading config from INI files with values encoded as JSON.

- *All* scanning methods can be overridden if some additional
  customization is required.

- A prescan method can be provided to handle values before the JSON
  scanners are applied.

- A fallback scanner method can be provided to handle additional types
  of values if none of the default scanners are suitable.

.. note:: See details below in :func:`decode` for using prescan or a
    fallback scanner.

Examples::

    >>> decode("+1")
    1
    >>> decode("1_000")
    1000
    >>> decode("[0b11]")
    [3]
    >>> d = decode("2021-06-24")
    >>> d.timetuple()[:5]
    (2021, 6, 24, 0, 0)
    >>> d.tzinfo
    tzlocal()
    >>> decode(" [1, 2 ,  3  ,   ] ")
    [1, 2, 3]
    >>> decode("[[]]")
    [[]]

"""
from configparser import ConfigParser, ExtendedInterpolation
from pathlib import Path
from typing import Any, Callable, Optional, TextIO, Tuple, Union

from . import scanner
from .exc import INIDecodeError
from .obj import JSONObject
from .scanner import Scanner, Scanner as Decoder


__all__ = ["decode", "decode_file", "Decoder"]


# Signature for prescanner and fallback scanner functions
AltScannerFunc = Callable[[Scanner, Callable, str, int], Tuple[Any, int]]


def decode(
    string: str,
    *,
    strict: bool = True,
    prescan: Optional[AltScannerFunc] = None,
    scan_object: Callable = scanner.scan_object,
    object_converter: Callable = JSONObject,
    scan_array: Callable = scanner.scan_array,
    scan_string: Callable = scanner.scan_string,
    scan_date: Callable = scanner.scan_date,
    scan_number: Callable = scanner.scan_number,
    fallback_scanner: Optional[AltScannerFunc] = None,
    enable_extras: bool = True,
    ignore_extra_data: bool = False,
    ini=False,
) -> Union[Any, Tuple[Any, int]]:
    """Scan JSONish string and return a Python object.

    - JSON object -> Python object (see below)
    - JSON array -> Python list
    - JSON string -> Python string
    - Literal date string (no quotes) -> Python datetime
    - JSON number -> Python number (int or float)
    - Empty string -> None

    By default, JSON objects are converted to simple Python namespace
    objects that allow attributes to be accessed via dotted or bracket
    notation. These objects can be converted to plain dicts with
    ``dict(obj)`` or you can use ``object_converter=None`` to get back
    plain dicts.

    A different ``object_converter`` can be passed to customize object
    creation, perhaps based on a type field::

        def converter(obj):
            if "__type__" in obj:
                # Convert to type based on __type__
                T = types[obj["__type__"]]
                return T(**obj)
            # Don't convert since no type was specified
            return obj

    When errors are encountered, various kinds of exceptions are
    thrown. These all derive from :class:`DecodeError`, which in turn
    derives from the builtin :class:`ValueError`.

    Examples::

        >>> import arrow, math

        >>> decode("") is None
        True

        >>> d = decode("2021-06-23")
        >>> d.timetuple()[:5]
        (2021, 6, 23, 0, 0)

        >>> t = arrow.now()
        >>> d = decode("12:00")
        >>> d.timetuple()[:5] == (t.year, t.month, t.day, 12, 0)
        True

        >>> d = decode("2021-06-23T12:00")
        >>> d.timetuple()[:6]
        (2021, 6, 23, 12, 0, 0)

        >>> decode("[inf, nan]")
        [inf, nan]

        >>> decode("E") == math.e
        True
        >>> (decode("π"), decode("PI")) == (math.pi, math.pi)
        True
        >>> (decode("τ"), decode("TAU")) == (math.tau, math.tau)
        True

        >>> decode("0"), decode("+0"), decode("-0"), decode("000")
        (0, 0, 0, 0)
        >>> decode("1"), decode("+1"), decode("-1")
        (1, 1, -1)

        >>> decode("1.0"), decode("+1.0"), decode("-1.0")
        (1.0, 1.0, -1.0)

        >>> decode("0b11"), decode("0o11"), decode("0x11")
        (3, 9, 17)

        >>> decode("{}", object_converter=None), decode("[]")
        ({}, [])

        >>> decode("[0b11, 11, 0x11]")
        [3, 11, 17]

    When the ``ignore_extra_data`` flag is set, a tuple will be returned
    containing 1) a Python object representing the part of the JSON
    string that was successfully parsed and 2) the index in the JSON
    string where the extra data starts. In most cases, extra data
    indicates an error, but this flag can be used to intentionally
    include extra data:

        >>> decode('{} # ignored', object_converter=None, ignore_extra_data=True)
        ({}, 3)

    There are a couple of advanced/esoteric/low level features for use
    where additional customization of parsing is required:

    - The prescanner. This is a callable that takes the scanner
      instance, the primary scan function, the JSON input string, and
      the current position; it can either return a value and the next
      position or ``None``. Returning ``None`` indicates that the
      prescanner didn't handle the string and that the regular scanners
      should be tried.

    - The fallback scanner. This is a callable that takes a scanner
      instance, the primary scan function, the JSON input string, and
      the current position; it must return a Python value along with
      the next position.

    """
    instance = scanner.Scanner(
        strict=strict,
        prescan=prescan,
        scan_object=scan_object,
        object_converter=object_converter,
        scan_array=scan_array,
        scan_string=scan_string,
        scan_date=scan_date,
        scan_number=scan_number,
        enable_extras=enable_extras,
        fallback_scanner=fallback_scanner,
    )
    if ini:
        return decode_ini(string, instance)
    return instance.decode(string, ignore_extra_data=ignore_extra_data)


def decode_ini(string, scanner):
    """Decode INI file with JSONish values.

    INI section and setting names are split on dots to create sub-dicts.
    For example::

        [section.one]
        a.b = 1

    will result in the following dict::

        {"section": "one": {"a": {"b": 1}}}

    To disable splitting a name, wrap it in parentheses. For example::

        [(section.one)]
        (a.b) = 1

    will result in the following dict::

        {"section.one": {"a.b": 1}}

    """
    result = {}
    parser = ConfigParser(interpolation=ExtendedInterpolation())
    parser.optionxform = lambda option: option
    parser.read_string(string)

    for section_name in parser.sections():
        obj = result
        section_path = parse_ini_name(section_name)
        for segment in section_path[:-1]:
            if segment not in obj:
                obj[segment] = {}
            obj = obj[segment]
        obj[section_path[-1]] = section_config = {}

        section_items = parser[section_name]
        for raw_name, raw_value in section_items.items():
            obj = section_config
            path = parse_ini_name(raw_name)
            for segment in path[:-1]:
                if segment not in obj:
                    obj[segment] = {}
                obj = obj[segment]
            value, _ = scanner.scan(raw_value)
            obj[path[-1]] = value

    return result


def parse_ini_name(name):
    """Parse section and setting names from INI file."""
    # NOTE: Groups can't be nested, so this is simplistic

    if "(" not in name and ")" not in name:
        return name.split(".")

    j = 0
    segments = []
    last_right_paren = 0

    while True:
        i = name.find("(", j)
        if i == -1:
            break

        unmatched_i = name.find(")", j)
        if -1 < unmatched_i < i:
            raise INIDecodeError(name, unmatched_i, f"Unmatched ) in name '{name}'")

        before = name[j:i]
        if before:
            if not before.endswith("."):
                raise INIDecodeError(name, i - 1, f"Expected dot in name '{name}'")
            segments.append(before[:-1])

        j = name.find(")", i + 1)
        if j == -1:
            raise INIDecodeError(name, i, f"Unmatched ( in name '{name}'")
        last_right_paren = j

        nested_i = name.find("(", i + 1)
        if -1 < nested_i < j:
            raise INIDecodeError(name, nested_i, f"Nested ( in name '{name}'")

        part = name[i + 1 : j]
        segments.append(part)

    unmatched_i = name.find(")", last_right_paren + 1)
    if unmatched_i != -1:
        raise INIDecodeError(name, unmatched_i, f"Unmatched ) in name '{name}'")

    after = name[j + 1 :]
    if after:
        if not after.startswith("."):
            raise INIDecodeError(name, j + 1, f"Expected dot in name '{name}'")
        segments.append(after[1:])

    return segments


def decode_file(
    file: Union[str, Path, TextIO],
    *,
    strict: bool = True,
    prescan: Optional[AltScannerFunc] = None,
    scan_object: Callable = scanner.scan_object,
    object_converter: Callable = JSONObject,
    scan_array: Callable = scanner.scan_array,
    scan_string: Callable = scanner.scan_string,
    scan_date: Callable = scanner.scan_date,
    scan_number: Callable = scanner.scan_number,
    fallback_scanner: Optional[AltScannerFunc] = None,
    enable_extras: bool = True,
    ignore_extra_data: bool = False,
    ini=None,
) -> Union[Any, Tuple[Any, int]]:
    """Read file, scan JSONish string, and return a Python object.

    This reads the file into a string, then calls :func:`decode`; see
    its docstring for details.

    TODO: This reads the whole file all at once. Maybe find a way to
          stream the file contents, although this would add a bit of
          complexity to the scanner in order to make it generically
          handle iterables of chars.

    """
    if isinstance(file, str):
        path = Path(file)
        with path.open() as fp:
            string = fp.read()
    elif isinstance(file, Path):
        path = file
        with file.open() as fp:
            string = fp.read()
    else:
        path = None
        string = file.read()
    if ini is None:
        if path is not None and path.suffix in (".cfg", ".ini", ".inijson"):
            ini = True
        else:
            ini = False
    return decode(
        string,
        strict=strict,
        prescan=prescan,
        scan_object=scan_object,
        object_converter=object_converter,
        scan_array=scan_array,
        scan_string=scan_string,
        scan_date=scan_date,
        scan_number=scan_number,
        fallback_scanner=fallback_scanner,
        enable_extras=enable_extras,
        ignore_extra_data=ignore_extra_data,
        ini=ini,
    )
